doc/source/ray-air/predictors.rst

```diff
@@ -143,10 +143,119 @@ Online Inference
-3. Optionally ``_predict_numpy`` for better performance when working with tensor data to avoid extra copies from Pandas conversions.
+If you're using an unsupported framework, or if built-in predictors are too inflexible,
+implement a custom predictor.
+
+How to implement a custom predictor
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To implement a custom :class:`~ray.train.predictor.Predictor`,
+subclass :class:`~ray.train.predictor.Predictor` and implement:
+
+* :meth:`~ray.train.predictor.Predictor.__init__`
+* :meth:`~ray.train.predictor.Predictor._predict_numpy` or :meth:`~ray.train.predictor.Predictor._predict_pandas`
+* :meth:`~ray.train.predictor.Predictor.from_checkpoint`
+
+.. tip::
+    You don't need to implement both
+    :meth:`~ray.train.predictor.Predictor._predict_numpy` and
+    :meth:`~ray.train.predictor.Predictor._predict_pandas`. Pick the method that's
+    easiest to implement. In general, if your model accepts a tensor as input, implement
+    :meth:`~ray.train.predictor.Predictor._predict_numpy`; otherwise, implement
+    :meth:`~ray.train.predictor.Predictor._predict_pandas`.
+
+Overview of required methods
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+__init__
+********
+
+Use the constructor to set instance attributes like ``self.model``.
+
+You must call the base class' constructor; otherwise, `Predictor.predict <ray.train.predictor.Predict.predict>` raises a
+``NotImplementedError``.
+
+.. testsetup::
+
+    from typing import Optional
+    from ray.data.preprocessor import Preprocessor
+    from ray.train.predictor import Predictor
+
+.. testcode::
+
+    class CustomPredictor(Predictor):
+        def __init__(self, preprocessor: Optional[Preprocessor] = None):
+            ...
+            super().__init__(self, preprocessor)
+
+
+_predict_numpy
+**************
+
+This method performs inference on a batch of NumPy data. It accepts a ``np.ndarray``
+or ``dict[str, np.ndarray]`` as input and returns a ``np.ndarray`` or
+``dict[str, np.ndarray]`` as output.
+
+The input type is determined by the type of :class:`~ray.data.Dataset` passed to
+:meth:`BatchPredictor.predict <ray.train.batch_predictor.BatchPredictor.predict>`:
+
+* If the input dataset is a tabular dataset, then the input type is ``dict[str, np.ndarray]``.
+* If the input dataset is a tensor dataset, then the input type is ``np.ndarray``.
+* If the input dataset is a simple dataset, then
+  :class:`~ray.train.batch_predictor.BatchPredictor` raises an error.
+
+Your output determines the type of dataset returned by
+:meth:`BatchPredictor.predict <ray.train.batch_predictor.BatchPredictor.predict>`:
+
+* If :meth:`~ray.train.predictor.Predictor._predict_numpy` returns a ``np.ndarray``, then :class:`~ray.train.batch_predictor.BatchPredictor` returns a tensor dataset
+* If :meth:`~ray.train.predictor.Predictor._predict_numpy` returns a ``dict[str, np.ndarray]``, then :class:`~ray.train.batch_predictor.BatchPredictor` returns a tabular dataset.
```

Comment:
Wasn't sure if we should include this information. On one hand, this isn't obvious and it might trip people up. On the other hand, it's hard to explain, and there's no user guide we can point users to for more information on simple/tensor/tabular datasets.

Reply:
I think we can change the language here a bit to just be something like "dict if dataset has column names, else just single numpy array".

Don't think we need to go into the details about tabular vs. tensor